### PR TITLE
New broadcasted gaussian integrals

### PR DESCRIPTION
**Pull request by ziofil**
### **User description**
Cleaned up and refactored Gaussian integrals, now working with batched, non batched and polynomial inputs

___

### **PR Type**
enhancement, tests

___

### **Description**
- Refactored Gaussian integral functions by removing deprecated functions and introducing `complex_gaussian_integral_1` and `complex_gaussian_integral_2`.
- Enhanced `join_Abc` function to support batch processing and improved input validation.
- Updated various modules to utilize the new Gaussian integral functions.
- Revised tests to cover new functionality and removed tests for deprecated functions.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>gaussian_integrals.py</strong><dd><code>Refactor Gaussian integral functions and enhance batch processing</code></dd></summary>
<hr>

mrmustard/physics/gaussian_integrals.py

<li>Removed <code>complex_gaussian_integral</code> and <code>contract_two_Abc</code> functions.<br> <li> Introduced <code>complex_gaussian_integral_1</code> and <code>complex_gaussian_integral_2</code> <br>functions.<br> <li> Enhanced <code>join_Abc</code> function to support batch processing.<br> <li> Improved input validation and error handling.<br>

</details>

  </td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/502/files#diff-a7b779ea109bae98fb3be0cb7ed5f6dba566d967f72666d24fbd01015f7962dc">+253/-266</a></td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>representations.py</strong><dd><code>Update Gaussian integral usage in representations</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

mrmustard/physics/representations.py

<li>Replaced <code>complex_gaussian_integral</code> with <code>complex_gaussian_integral_1</code>.<br> <li> Updated matrix multiplication to use <code>complex_gaussian_integral_2</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/502/files#diff-f9144a4d7f566d31653dfdc366d79fb562073ce83ada8346055dd63405075ff0">+9/-17</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>triples.py</strong><dd><code>Update triples to use new Gaussian integral function</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

mrmustard/physics/triples.py

- Replaced `contract_two_Abc` with `complex_gaussian_integral_2`.

</details>

  </td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/502/files#diff-615d7fb7ff981fd55f46b6a2e69b1f999b052812ff6d2d49c00ba9f0a98b0004">+5/-7</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_circuit_components_utils.py</strong><dd><code>Update circuit components utils tests for new integrals</code>&nbsp; &nbsp; </dd></summary>
<hr>

tests/test_lab_dev/test_circuit_components_utils.py

<li>Updated tests to use <code>complex_gaussian_integral_1</code> and <br><code>complex_gaussian_integral_2</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/502/files#diff-bbae102edbdb2a5d99dcf4ca76d04d6c1e391adf176307285b3ae466afb5ad5f">+7/-7</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>test_gaussian_integrals.py</strong><dd><code>Add tests for new Gaussian integral functions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/test_physics/test_gaussian_integrals.py

<li>Added tests for <code>complex_gaussian_integral_1</code> and <br><code>complex_gaussian_integral_2</code>.<br> <li> Removed tests for deprecated functions.<br>

</details>

  </td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/502/files#diff-f64960aa126fc9053f056b46e568e9a1cd22ccb908dbccad294f7d67a1fbbb1b">+133/-72</a></td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>test_representations.py</strong><dd><code>Update representation tests for new integrals</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/test_physics/test_representations.py

<li>Updated tests to use <code>complex_gaussian_integral_1</code> and <br><code>complex_gaussian_integral_2</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/502/files#diff-54974d201eb9bf8f53ddfcbb797fab5a3076b6a533d10fa60228d8395ab409a4">+4/-4</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

File: mrmustard/physics/gaussian_integrals.py [hunk 1]
```diff
@@ -259,216 +160,307 @@ def reorder_abc(Abc: tuple, order: Sequence[int]):
...
 
-def contract_two_Abc(
-    Abc1: tuple[ComplexMatrix, ComplexVector, ComplexTensor],
-    Abc2: tuple[ComplexMatrix, ComplexVector, ComplexTensor],
-    idx1: Sequence[int],
-    idx2: Sequence[int],
-):
-    r"""
-    Returns the contraction of two ``(A,b,c)`` triples with given indices.
+def join_Abc(Abc1, Abc2, mode="kron"):  # pylint: disable=too-many-statements
```
Comment (apchytr) [inline mrmustard/physics/gaussian_integrals.py:182]: Can we include type hints?

File: mrmustard/physics/gaussian_integrals.py [hunk 2]
```diff
@@ -78,115 +78,16 @@ def real_gaussian_integral(
...
-
 def join_Abc_real(
     Abc1: tuple[ComplexMatrix, ComplexVector, ComplexTensor],
     Abc2: tuple[ComplexMatrix, ComplexVector, ComplexTensor],
     idx1: Sequence[int],
     idx2: Sequence[int],
 ):
-    r"""Direct sum of two ``(A,b,c)`` triples into a single ``(A,b,c)`` triple, where indices corresponding to the same variable are "fused together", by considering their Bargmann function has having the same variables. For example ``idx1=(0,1,2)`` and ``idx2=(1,2,3)`` means that indices 1 and 2 will be fused because they are present on both tuples. This is useful for computing real Gaussian integrals where the variable on either object is the same, rather than a pair of conjugate variables for complex Gaussian integrals.
+    r"""Direct sum of two ``(A,b,c)`` triples into a single ``(A,b,c)`` triple, where indices corresponding to the same variable are "fused together",
+    by considering their Bargmann function has having the same variables. For example ``idx1=(0,1,2)`` and ``idx2=(1,2,3)`` means that indices 1 and 2
```
Comment (apchytr) [inline mrmustard/physics/gaussian_integrals.py:88]:
```suggestion
    by considering their Bargmann function as having the same variables. For example ``idx1=(0,1,2)`` and ``idx2=(1,2,3)`` means that indices 1 and 2
```

File: mrmustard/physics/gaussian_integrals.py [hunk 3]
```diff
@@ -259,216 +160,307 @@ def reorder_abc(Abc: tuple, order: Sequence[int]):
...
-    r"""
-    Returns the contraction of two ``(A,b,c)`` triples with given indices.
+def join_Abc(Abc1, Abc2, mode="kron"):  # pylint: disable=too-many-statements
+    r"""Joins two ``(A,b,c)`` triples into a single ``(A,b,c)``.
 
-    Note that the indices must be a complex variable pairs with each other to make this contraction meaningful. Please make sure
-    the corresponding complex variable with respect to your Abc triples.
-    For examples, if the indices of Abc1 denotes the variables ``(\alpha, \beta)``, the indices of Abc2 denotes the variables
-    ``(\alpha^*,\gamma)``, the contraction only works with ``idx1 = [0], idx2 = [0]``.
+    It support including a batch dimension, e.g. ``A1.shape = (batch, n1, n1)``,
```
Comment (apchytr) [inline mrmustard/physics/gaussian_integrals.py:185]:
```suggestion
    It supports including a batch dimension, e.g. ``A1.shape = (batch, n1, n1)``,
```

File: mrmustard/physics/gaussian_integrals.py [hunk 4]
```diff
@@ -259,216 +160,307 @@ def reorder_abc(Abc: tuple, order: Sequence[int]):
...
+    batched2 = len(A2.shape) == 3
+
+    if not batched1:
+        A1 = math.atleast_3d(A1)
+        b1 = math.atleast_2d(b1)
+        c1 = math.atleast_1d(c1)
+    if not batched2:
+        A2 = math.atleast_3d(A2)
+        b2 = math.atleast_2d(b2)
+        c2 = math.atleast_1d(c2)
```
Comment (apchytr) [inline mrmustard/physics/gaussian_integrals.py:232]: Can't we just always call these? Then you can avoid calling astensor above

File: mrmustard/physics/gaussian_integrals.py [hunk 5]
```diff
@@ -259,216 +160,307 @@ def reorder_abc(Abc: tuple, order: Sequence[int]):
...
 
+    # 4. return according to inputs
 
-def complex_gaussian_integral_2(
-    Abc1: tuple, Abc2: tuple, idx1: tuple[int, ...], idx2: tuple[int, ...], measure: float = -1
-):
-    r"""Computes the Gaussian integral of the exponential of a complex quadratic form.
-    The integral is defined as (note that in general we integrate over a subset of 2m dimensions):
+    if not batched1 and not batched2:
+        return A[0], b[0], c[0]
```
Comment (apchytr) [inline mrmustard/physics/gaussian_integrals.py:323]: I dont fully understand why we wouldn't we just always return batched? We really should stick to having all internal methods expect `(A,b,c)` to be batched and return things batched for consistency no? With the exceptions being methods with a `batched` parameter exposed to the user for convenience

File: mrmustard/physics/gaussian_integrals.py [hunk 6]
```diff
@@ -259,216 +160,307 @@ def reorder_abc(Abc: tuple, order: Sequence[int]):
...
-        D = math.gather(math.gather(A, idx, axis=-1), not_idx, axis=-2)
-        R = math.gather(math.gather(A, not_idx, axis=-1), not_idx, axis=-2)
-        bR = math.gather(b, not_idx, axis=-1)
-        A_post = R - math.matmul(D, math.inv(M), math.transpose(D))
-        b_post = bR - math.matvec(D, math.solve(M, bM))
+    determinant = math.det(M)
+    D = math.gather(math.gather(A, idx, axis=-1), not_idx, axis=-2)
+    R = math.gather(math.gather(A, not_idx, axis=-1), not_idx, axis=-2)
+    bR = math.gather(b, not_idx, axis=-1)
+    if np.all(math.abs(determinant)) > 1e-12:
```
Comment (apchytr) [inline mrmustard/physics/gaussian_integrals.py:410]: Should this be a `math.allclose`?

File: mrmustard/physics/gaussian_integrals.py [hunk 7]
```diff
@@ -259,216 +160,307 @@ def reorder_abc(Abc: tuple, order: Sequence[int]):
...
-    Abc1: tuple[ComplexMatrix, ComplexVector, ComplexTensor],
-    Abc2: tuple[ComplexMatrix, ComplexVector, ComplexTensor],
-):
-    r"""Joins two ``(A,b,c)`` triples into a single ``(A,b,c)`` triple by block addition of the ``A``
-    matrices and concatenating the ``b`` vectors.
-
-    Arguments:
-        Abc1: the first ``(A,b,c)`` triple
-        Abc2: the second ``(A,b,c)`` triple
+def complex_gaussian_integral_2(Abc1, Abc2, idx1, idx2, measure=-1, mode: str = "kron"):
```
Comment (apchytr) [inline mrmustard/physics/gaussian_integrals.py:427]: Can you include type hints?

File: mrmustard/physics/representations.py [hunk 8]
```diff
@@ -436,12 +436,7 @@ def trace(self, idx_z: tuple[int, ...], idx_zconj: tuple[int, ...]) -> Bargmann:
         Returns:
             Bargmann: the ansatz with the given indices traced over
         """
-        A, b, c = [], [], []
-        for Abc in zip(self.A, self.b, self.c):
-            Aij, bij, cij = complex_gaussian_integral(Abc, idx_z, idx_zconj, measure=-1.0)
-            A.append(Aij)
-            b.append(bij)
-            c.append(cij)
+        A, b, c = complex_gaussian_integral_1(self.triple, idx_z, idx_zconj, measure=-1.0)
```
Comment (apchytr) [inline mrmustard/physics/representations.py:439]: Love seeing things simplified like this 😄

File: mrmustard/physics/gaussian_integrals.py (hunk 6 above)
Comment (ziofil) [reply mrmustard/physics/gaussian_integrals.py]: no, I want to test if they are all not zero

**Review by apchytr — APPROVED** [review]

**Review by apchytr — APPROVED** [review]

**Pull request merged**